x/earn/keeper/grpc_query.go

```diff
@@ -33,3 +39,281 @@ func (s queryServer) Params(ctx context.Context, req *types.QueryParamsRequest)
+		}
+
+		value, err := getAccountValue(sdkCtx, s.keeper, owner, shareRecord.AmountSupplied)
+		if err != nil {
+			return nil, status.Error(codes.InvalidArgument, err.Error())
+		}
+
+		return &types.QueryDepositsResponse{
+			Deposits: []types.DepositResponse{
+				{
+					Depositor:      owner.String(),
+					AmountSupplied: shareRecord.AmountSupplied,
+					Value:          value,
+				},
+			},
+			Pagination: nil,
+		}, nil
+	}
+
+	// 2. All accounts, specific vault
+	if req.Owner == "" && req.Denom != "" {
+		_, found := s.keeper.GetVaultRecord(sdkCtx, req.Denom)
+		if !found {
+			return nil, status.Error(codes.NotFound, "Vault record for denom not found")
+		}
+
+		deposits := []types.DepositResponse{}
+		store := prefix.NewStore(sdkCtx.KVStore(s.keeper.key), types.VaultShareRecordKeyPrefix)
+
+		pageRes, err := query.FilteredPaginate(
+			store,
+			req.Pagination,
+			func(key []byte, value []byte, accumulate bool) (bool, error) {
+				var record types.VaultShareRecord
+				err := s.keeper.cdc.Unmarshal(value, &record)
+				if err != nil {
+					return false, err
+				}
+
+				// Only those that have amount of requested denom
+				if record.AmountSupplied.AmountOf(req.Denom).IsZero() {
+					// inform paginate that there was no match on this key
+					return false, nil
+				}
+
+				if accumulate {
+					accValue, err := getAccountValue(sdkCtx, s.keeper, record.Depositor, record.AmountSupplied)
+					if err != nil {
+						return false, err
+					}
+
+					// only add to results if paginate tells us to
+					deposits = append(deposits, types.DepositResponse{
+						Depositor:      record.Depositor.String(),
+						AmountSupplied: record.AmountSupplied,
+						Value:          accValue,
+					})
+				}
+
+				// inform paginate that were was a match on this key
+				return true, nil
+			},
+		)
+
+		if err != nil {
+			return nil, err
+		}
+
+		return &types.QueryDepositsResponse{
+			Deposits:   deposits,
+			Pagination: pageRes,
+		}, nil
+	}
+
+	// 3. Specific account, all vaults
+	if req.Owner != "" && req.Denom == "" {
+		owner, err := sdk.AccAddressFromBech32(req.Owner)
+		if err != nil {
+			return nil, status.Error(codes.InvalidArgument, "Invalid address")
+		}
+
+		deposits := []types.DepositResponse{}
+
+		accountShare, found := s.keeper.GetVaultShareRecord(sdkCtx, owner)
+		if !found {
+			return nil, status.Error(codes.NotFound, "No deposit found for owner")
+		}
+
+		value, err := getAccountValue(sdkCtx, s.keeper, owner, accountShare.AmountSupplied)
+		if err != nil {
+			return nil, status.Error(codes.InvalidArgument, err.Error())
+		}
+
+		deposits = append(deposits, types.DepositResponse{
+			Depositor:      owner.String(),
+			AmountSupplied: accountShare.AmountSupplied,
+			Value:          value,
+		})
+
+		return &types.QueryDepositsResponse{
+			Deposits:   deposits,
+			Pagination: nil,
+		}, nil
+	}
+
+	// 4. All accounts, all vaults
```

Comment:
suggestion: I've found when I want to add comments dividing up a function into sections, it usually means the function is too big and doing things at several layers of abstraction. The sections can often be pulled out into a functions with similar names to the comments.

Reply:
Will update in follow up refactor